### PR TITLE
[Typed throws] Implement thrown type inference for do..catch within closures

### PR DESCRIPTION
**Pull request by DougGregor**
Start classifying all potential throw sites within a constraint
system and associate them with the nearest enclosing catch node. Then,
determine the thrown error type for a given catch node by taking the
union of the thrown errors at each potential throw site. Use this to
compute the error type thrown from the body of a `do..catch` block
within a closure.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: include/swift/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -1547,6 +1575,13 @@ class Solution {
   llvm::MapVector<const CaseLabelItem *, CaseLabelItemInfo>
       caseLabelItems;
 
+  /// Maps catch nodes to the set of potential throw sites that will be caught
+  /// at that location.
+
+  /// The set of opened types for a given locator.
+  std::vector<std::pair<CatchNode, PotentialThrowSite>>
+      potentialThrowSites;
```
Comment (xedin) [inline include/swift/Sema/ConstraintSystem.h:1583]: Nit: please add the size to `Solution::getTotalMemory()`

**Review by xedin — APPROVED** [review]

File: include/swift/Sema/ConstraintSystem.h (hunk 1 above)
Comment (DougGregor) [reply include/swift/Sema/ConstraintSystem.h:1583]: Sure, I'll do this when I rework the data structure here. Thank you!

**Pull request merged**